ui/src/app/controllers/views/ControllerDetails/ControllerConfiguration/ControllerPowerConfiguration.tsx

```diff
@@ -0,0 +1,164 @@
+  const canEdit = useCanEdit(controller);
+  const isDetails = isControllerDetails(controller);
+  const initialPowerParameters = useInitialPowerParameters(
+    (isDetails && controller.power_parameters) || {}
+  );
+  const isAdmin = useSelector(authSelectors.isAdmin);
+
+  const powerParametersSchema =
+    generatePowerParametersSchema(selectedPowerType);
+
+  const PowerFormSchema: SchemaOf<PowerFormValues> = Yup.object()
+    .shape({
+      powerParameters: Yup.object().shape(powerParametersSchema),
+      powerType: Yup.string().required("Power type is required"),
+    })
+    .defined();
+
+  useEffect(() => {
+    // We set the selected power type outside the scope of the form as its used
+    // to generate the form's validation schema.
+    if (controller?.power_type) {
+      const powerType = getPowerTypeFromName(powerTypes, controller.power_type);
+      setSelectedPowerType(powerType);
+    }
+  }, [controller?.power_type, powerTypes]);
+
+  if (!isDetails || powerTypesLoading) {
+    return <Spinner text="Loading..." />;
+  }
+
+  return (
+    <EditableSection
+      canEdit={canEdit}
+      hasSidebarTitle
+      renderContent={(editing, setEditing) =>
+        editing ? (
+          <FormikForm
+            aria-label={Label.Title}
+            allowAllEmpty
+            allowUnchanged
+            cleanup={cleanup}
+            editable={editing}
+            errors={errors}
+            initialValues={{
+              powerType: controller.power_type,
+              powerParameters: initialPowerParameters,
+            }}
+            onSaveAnalytics={{
+              action: "Configure power",
+              category: "Machine details",
+              label: "Save changes",
+            }}
+            onCancel={() => setEditing(false)}
+            onSubmit={(values) => {
+              const params = {
+                power_parameters: formatPowerParameters(
+                  selectedPowerType,
+                  values.powerParameters
+                ),
+                power_type: values.powerType,
+                system_id: controller.system_id,
+              };
+              dispatch(controllerActions.update(params));
+            }}
+            onSuccess={() => setEditing(false)}
+            onValuesChanged={(values) => {
+              const powerType = getPowerTypeFromName(
+                powerTypes,
+                values.powerType
+              );
+              setSelectedPowerType(powerType);
+            }}
+            saved={saved}
+            saving={saving}
+            submitLabel="Save changes"
+            validationSchema={PowerFormSchema}
+          >
+            {controller.power_bmc_node_count > 1 ? (
+              <Notification severity="information">
+                This power controller manages{" "}
+                {controller?.power_bmc_node_count - 1}
+                other {controller.power_bmc_node_count > 3 ? "nodes" : "node"}
```

Comment:
This was correct, but slightly confusing (`power_bmc_node_count` includes the controller itself), so controller with 2 other nodes has `power_bmc_node_count = 3`. I refactored it so that it hopefully is more clear.

Reply:
Ah I see, thanks